codec/src/main/java/io/netty/handler/codec/compression/CompressionHandler.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2021 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.compression;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.Promise;
+
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+/**
+ * {@link ChannelHandler} which uses a {@link Compressor} for compressing the written {@link ByteBuf}s.
+ */
+public class CompressionHandler implements ChannelHandler {
+
+    private final Supplier<? extends Compressor> compressorSupplier;
+    private Compressor compressor;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param compressorSupplier  the {@link Supplier} that is used to create the {@link Compressor}.
+     */
+    public CompressionHandler(Supplier<? extends Compressor> compressorSupplier) {
+        this.compressorSupplier = Objects.requireNonNull(compressorSupplier, "compressorSupplier");
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        compressor = compressorSupplier.get();
+    }
+
+    @Override
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+        if (compressor != null) {
+            finish(ctx, false);
+            compressor.close();
+            compressor = null;
+        }
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        if (compressor != null) {
+            compressor.close();
+            compressor = null;
```

Comment:
This means if `channelInactive()` happens before `handlerRemoved()` then we won't call `finish()`, is this intentional? Can you explain and perhaps document the reasoning?

Reply:
yes it is... `finish()` only make sense if you are the one that actually close the channel.